File: SOLID-Java/src/main/java/singleresponsibility/after/XMLReaderWriter.java
package singleresponsibility.after;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.Collectors;

public class XMLReaderWriter
{
	private final static Logger LOGGER = Logger.getLogger(XMLReaderWriter.class.getName());
	
	public List<String> read()
	{
		List<String> fileNames = new ArrayList<>();
		LOGGER.info("Reading XMLs");
		final URL resource = getClass().getResource("/xmls");
		final File file = new File(resource.getPath());
		if (file.exists() && file.isDirectory())
		{
			final List<File> files = Arrays.asList(file.listFiles());
			files.sort((File o1, File o2) -> o1.getName().compareTo(o2.getName()));
			fileNames = files.stream().map(x -> x.getName()).collect(Collectors.toList());
		}
		return fileNames;
	}

	public void write() throws Exception
	{
		FileWriter writer = null;
		try
		{
			final File file = new File("src/main/resources/xmls/MyXML.xml");
			if (file.createNewFile())
			{
				LOGGER.info("File is created!");
			} else
			{
				throw new Exception("File already exists.");
			}

			writer = new FileWriter(file);
			writer.write("Test xml");
		} catch (IOException e)
		{
			LOGGER.log(Level.SEVERE, e.getMessage());
		}
		finally {
			writer.close();
		}
	}
}

File: SOLID-Scala/src/main/scala/openclose/readwriter/XMLReaderWriter.scala
package openclose.readwriter

import java.io.File
import java.io.IOException
import java.util.logging.Level
import java.io.FileWriter
import java.util.logging.Level
import java.util.logging.Logger
import scala.util.control.NonFatal
import scala.util.Try

class XMLReaderWriter extends FileReaderWriter {
  val LOGGER = Logger.getLogger(this.getClass.getName)

  def read(): List[String] = {
    val fileNames = {
      LOGGER.info("Reading XMLs")
      val resource = getClass.getResource("/xmls")
      val file = new File(resource.getPath())
      if (file.exists() && file.isDirectory()) {
        val files = file.listFiles().sortBy { x => x.getName }
        files.map(x => x.getName).toList
      } else {
        List.empty
      }
    }
    fileNames
  }

  def write(fileName: String): Unit = {
    var writer: FileWriter = null
    val path = "src/main/resources/"
    try {
      val file = new File(path + s"xmls/${fileName}.xml")
      if (file.createNewFile()) {
        LOGGER.info("File is created!")
      } else {
        LOGGER.info("File already exists.")
      }
      writer = new FileWriter(file)
      writer.write("Test xml")
    } catch {
      case NonFatal(e) => LOGGER.log(Level.SEVERE, e.getMessage())
    } finally {
      if (writer != null) {
        writer.close()
      }
    }
  }
  
}